AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/SQLite/SQLStatement+Select.swift

```diff
@@ -101,12 +72,81 @@ struct SelectStatement: SQLStatement {
             \(paginationInput.sqlStatement)
             """
         }
+        return SelectStatementMetadata(statement: sql,
+                                       columnMapping: columnMapping,
+                                       bindings: bindings)
+    }
+
+    /// Tuple that represents a pair of `columns` and `statements`
+    private typealias JoinStatements = ([String], [String], ColumnMapping)
+
+    /// Walk through the associations recursively to generate join statements.
+    ///
+    /// Implementation note: this should be revisited once we define support
+    /// for explicit `eager` vs `lazy` associations.
+    private static func joins(from schema: ModelSchema) -> JoinStatements {
+        var columns: [String] = []
+        var joinStatements: [String] = []
+        var columnMapping: ColumnMapping = [:]
+
+        func visitAssociations(node: ModelSchema, namespace: String = "root") {
+            for foreignKey in node.foreignKeys {
+                let associatedModelType = foreignKey.requiredAssociatedModel
+                let associatedSchema = associatedModelType.schema
+                let associatedTableName = associatedModelType.schema.name
+
+                // columns
+                let alias = namespace == "root" ? foreignKey.name : "\(namespace).\(foreignKey.name)"
+                let associatedColumn = associatedSchema.primaryKey.columnName(forNamespace: alias)
+                let foreignKeyName = foreignKey.columnName(forNamespace: namespace)
+
+                // append columns from relationships
+                columns += associatedSchema.columns.map { field -> String in
+                    columnMapping.updateValue((associatedSchema, field), forKey: "\(alias).\(field.name)")
+                    return field.columnName(forNamespace: alias) + " as " + field.columnAlias(forNamespace: alias)
+                }
+
+                let joinType = foreignKey.isRequired ? "inner" : "left outer"
+
+                joinStatements.append("""
+                \(joinType) join \(associatedTableName) as "\(alias)"
+                  on \(associatedColumn) = \(foreignKeyName)
+                """)
+                visitAssociations(node: associatedModelType.schema,
+                                  namespace: alias)
+            }
+        }
+        visitAssociations(node: schema)
+
+        return (columns, joinStatements, columnMapping)
+    }
+
+}
+
+/// Represents a `select` SQL statement associated with a `Model` instance and
+/// optionally composed by a `ConditionStatement`.
+struct SelectStatement: SQLStatement {
+
+    let modelType: Model.Type
+    let metadata: SelectStatementMetadata
+
+    init(from modelType: Model.Type,
+         predicate: QueryPredicate? = nil,
+         paginationInput: QueryPaginationInput? = nil,
+         additionalStatements: String? = nil) {
+        self.modelType = modelType
+        self.metadata = .get(from: modelType,
```

Comment:
A lot happens here in the init method, and I think this code path is sync with the developers code `Amplify.DataStore.query`. Is this an expensive statement that can affect the query api?

In future we should add some performance test around this.

Reply:
+1 to performance benchmarks